File: benchmarks/java/CpuBenchmarkFair.java
// CPU Benchmark - Fair (uses built-in sort)
import java.util.Arrays;
import java.util.Collections;

public class CpuBenchmarkFair {
    static long fibRecursive(long n) { return n <= 1 ? n : fibRecursive(n-1) + fibRecursive(n-2); }

    static double benchmarkFibonacci() {
        long start = System.currentTimeMillis();
        long sum = 0; for (int i = 0; i < 40; i++) sum += fibRecursive(i);
        long elapsed = System.currentTimeMillis() - start;
        System.out.println("Fibonacci(40): " + elapsed + " ms");
        return elapsed;
    }

    static double benchmarkMatrixMultiply() {
        int n = 512;
        double[][] a = new double[n][n], b = new double[n][n], c = new double[n][n];
        for (int i = 0; i < n; i++) for (int j = 0; j < n; j++) { a[i][j] = i*j; b[i][j] = i+j; }
        long start = System.currentTimeMillis();
        for (int i = 0; i < n; i++) for (int j = 0; j < n; j++) { double s = 0; for (int k = 0; k < n; k++) s += a[i][k]*b[k][j]; c[i][j] = s; }
        long elapsed = System.currentTimeMillis() - start;
        System.out.println("Matrix Multiply (512x512): " + elapsed + " ms");
        return elapsed;
    }

    static double benchmarkSorting() {
        int n = 1_000_000;
        long[] arr = new long[n]; for (int i = 0; i < n; i++) arr[i] = n - i - 1;
        long start = System.currentTimeMillis();
        Arrays.sort(arr);
        long elapsed = System.currentTimeMillis() - start;
        System.out.println("Sorting (1M elements): " + elapsed + " ms");
        return elapsed;
    }

    static double benchmarkPrimeSieve() {
        int limit = 10_000_000;
        boolean[] sieve = new boolean[limit + 1]; Arrays.fill(sieve, true); sieve[0] = sieve[1] = false;
        long start = System.currentTimeMillis();
        for (int p = 2; p*p <= limit; p++) if (sieve[p]) for (int i = p*p; i <= limit; i += p) sieve[i] = false;
        int count = 0; for (boolean b : sieve) if (b) count++;
        long elapsed = System.currentTimeMillis() - start;
        System.out.println("Prime Sieve (" + limit + "): " + count + " primes in " + elapsed + " ms");
        return elapsed;
    }

    static double benchmarkStringProcessing() {
        int n = 100_000;
        long start = System.currentTimeMillis();
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) sb.append("Item ").append(i).append(" ");
        long elapsed = System.currentTimeMillis() - start;
        System.out.println("String Concatenation (" + n + "): " + elapsed + " ms");
        return elapsed;
    }

    public static void main(String[] args) {
        System.out.println("=== Java CPU Benchmark ===\n");
        double total = 0;
        total += benchmarkFibonacci(); total += benchmarkMatrixMultiply();
        total += benchmarkSorting(); total += benchmarkPrimeSieve();
        total += benchmarkStringProcessing();
        System.out.println("\nTotal: " + total + " ms");
    }
}

File: benchmarks/rust/cpu_benchmark_fair.rs
// CPU Benchmark - Fair (uses built-in sort, matches real-world usage)
use std::time::Instant;

fn fib_recursive(n: u64) -> u64 {
    if n <= 1 { return n; }
    fib_recursive(n - 1) + fib_recursive(n - 2)
}

fn benchmark_fibonacci() -> f64 {
    let start = Instant::now();
    let mut sum = 0u64;
    for i in 0..40 { sum += fib_recursive(i); }
    std::hint::black_box(sum);
    let elapsed = start.elapsed().as_millis() as f64;
    println!("Fibonacci(40): {} ms", elapsed);
    elapsed
}

fn benchmark_matrix_multiply() -> f64 {
    let n = 512;
    let mut a = vec![vec![0.0f64; n]; n];
    let mut b = vec![vec![0.0f64; n]; n];
    let mut c = vec![vec![0.0f64; n]; n];
    for i in 0..n { for j in 0..n { a[i][j] = (i*j) as f64; b[i][j] = (i+j) as f64; } }
    let start = Instant::now();
    for i in 0..n { for j in 0..n { let mut s = 0.0; for k in 0..n { s += a[i][k] * b[k][j]; } c[i][j] = s; } }
    std::hint::black_box(c[0][0]);
    let elapsed = start.elapsed().as_millis() as f64;
    println!("Matrix Multiply (512x512): {} ms", elapsed);
    elapsed
}

fn benchmark_sorting() -> f64 {
    let n = 1_000_000i64;
    let mut arr: Vec<i64> = (0..n).rev().collect();
    let start = Instant::now();
    arr.sort_unstable();
    std::hint::black_box(arr[0]);
    let elapsed = start.elapsed().as_millis() as f64;
    println!("Sorting (1M elements): {} ms", elapsed);
    elapsed
}

fn benchmark_prime_sieve() -> f64 {
    let limit = 10_000_000usize;
    let mut sieve = vec![true; limit + 1];
    sieve[0] = false; sieve[1] = false;
    let start = Instant::now();
    let mut p = 2;
    while p * p <= limit { if sieve[p] { let mut i = p * p; while i <= limit { sieve[i] = false; i += p; } } p += 1; }
    let count = sieve.iter().filter(|&&x| x).count();
    let elapsed = start.elapsed().as_millis() as f64;
    println!("Prime Sieve ({}): {} primes in {} ms", limit, count, elapsed);
    elapsed
}

fn benchmark_string_processing() -> f64 {
    let n = 100_000usize;
    let start = Instant::now();
    let mut result = String::new();
    for i in 0..n { result.push_str(&format!("Item {}: ", i)); }
    std::hint::black_box(result.len());
    let elapsed = start.elapsed().as_millis() as f64;
    println!("String Concatenation ({}): {} ms", n, elapsed);
    elapsed
}

fn main() {
    println!("=== Rust CPU Benchmark ===\n");
    let mut total = 0.0;
    total += benchmark_fibonacci();
    total += benchmark_matrix_multiply();
    total += benchmark_sorting();
    total += benchmark_prime_sieve();
    total += benchmark_string_processing();
    println!("\nTotal: {} ms", total);
}
